homeassistant/components/upnp/__init__.py

```diff
@@ -0,0 +1,169 @@
+"""
+Will open a port in your router for Home Assistant and provide statistics.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/upnp/
+"""
+import asyncio
+from ipaddress import ip_address
+
+import aiohttp
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import dispatcher
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.components.discovery import DOMAIN as DISCOVERY_DOMAIN
+
+from .const import (
+    CONF_ENABLE_PORT_MAPPING, CONF_ENABLE_SENSORS,
+    CONF_HASS, CONF_LOCAL_IP, CONF_PORTS,
+    CONF_UDN, CONF_SSDP_DESCRIPTION,
+    SIGNAL_REMOVE_SENSOR,
+)
+from .const import DOMAIN
+from .const import LOGGER as _LOGGER
+from .config_flow import ensure_domain_data
+from .device import Device
+
+
+REQUIREMENTS = ['async-upnp-client==0.12.4']
+DEPENDENCIES = ['http']
+
+NOTIFICATION_ID = 'upnp_notification'
+NOTIFICATION_TITLE = 'UPnP/IGD Setup'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_ENABLE_PORT_MAPPING, default=False): cv.boolean,
+        vol.Optional(CONF_ENABLE_SENSORS, default=True): cv.boolean,
+        vol.Optional(CONF_LOCAL_IP): vol.All(ip_address, cv.string),
+        vol.Optional(CONF_PORTS):
+            vol.Schema({
+                vol.Any(CONF_HASS, cv.positive_int):
+                    vol.Any(CONF_HASS, cv.positive_int)
+            })
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def _substitute_hass_ports(ports, hass_port):
+    """Substitute 'hass' for the hass_port."""
+    ports = ports.copy()
+
+    # substitute 'hass' for hass_port, both keys and values
+    if CONF_HASS in ports:
+        ports[hass_port] = ports[CONF_HASS]
+        del ports[CONF_HASS]
+
+    for port in ports:
+        if ports[port] == CONF_HASS:
+            ports[port] = hass_port
+
+    return ports
+
+
+# config
+async def async_setup(hass: HomeAssistantType, config: ConfigType):
+    """Register a port mapping for Home Assistant via UPnP."""
+    ensure_domain_data(hass)
+
+    # ensure sane config
+    if DOMAIN not in config:
+        return True
+
+    if DISCOVERY_DOMAIN not in config:
```

Comment:
This makes no sense, we should just add it as a dependency?

Reply:
When running tests/tox and `discovery` is in `DEPENDENCIES`, I get errors such as:
```
Traceback (most recent call last):
  File "/home/steven/src/ha/home-assistant/homeassistant/setup.py", line 145, in _async_setup_component
    hass, processed_config)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/discovery.py", line 112, in async_setup
    from netdisco.discovery import NetworkDiscovery
ModuleNotFoundError: No module named 'netdisco'
```

I must have made this as a quick work-around and never thought about it anymore.

I see the `konnected` component depends on `discovery` as well, but has no tests.

What is the proposed fix, @balloob?

Reply:
Use `MockDependency` from tests.common

Reply:
@balloob How should `MockDependency` be used? If I try something like:
```
async def test_async_setup_auto_config(hass: HomeAssistantType):
    """Test async_setup."""
    # setup component, enable auto_config
    config = {'upnp': {}, 'discovery': {}}
    with MockDependency('discovery'):
        await async_setup_component(hass, 'upnp', config)
```

I always end up with something like:
```
INFO:homeassistant.loader:Loaded upnp from homeassistant.components.upnp
INFO:homeassistant.loader:Loaded http from homeassistant.components.http
INFO:homeassistant.loader:Loaded discovery from homeassistant.components.discovery
ERROR:homeassistant.config:Invalid config for [discovery]: required key not provided @ data['discovery']. Got None. (See ?, line ?). Please check the docs at https://home-assistant.io/components/discovery/
ERROR:homeassistant.setup:Setup failed for discovery: Invalid config.
INFO:homeassistant.setup:Setting up http
INFO:homeassistant.setup:Setup of domain http took 0.0 seconds.
DEBUG:homeassistant.core:Bus:Handling <Event component_loaded[L]: component=http>
DEBUG:homeassistant.core:Bus:Handling <Event call_service[L]: domain=persistent_notification, service=create, service_data=title=Invalid config, message=The following components and platforms could not be set up:

 - [discovery](https://home-assistant.io/components/discovery/)

Please check your config., notification_id=invalid_config, service_call_id=0382331dfc8b43d183b72f3710aae924>
DEBUG:homeassistant.core:Bus:Handling <Event call_service[L]: domain=persistent_notification, service=create, service_data=title=Invalid config, message=The following components and platforms could not be set up:

 - [discovery](https://home-assistant.io/components/discovery/)

Please check your config., notification_id=invalid_config, service_call_id=7ba98f6978434fd9af098b0373193d67>
ERROR:homeassistant.setup:Unable to set up dependencies of upnp. Setup failed for dependencies: discovery
ERROR:homeassistant.setup:Setup failed for upnp: Could not set up all dependencies.
```

Using `MockDependency` as a decorator does not work for async functions, but `tests/components/cast/test_init.py` uses it with `with` as well.

Reply:
you need to stay in the with block until the config flow is done initializing.